src/main/java/pl/allegro/tech/opel/EvalContext.java

```diff
@@ -20,7 +20,7 @@ static EvalContext empty() {
     class Builder {
         private final Map<String, CompletableFuture<Object>> variables = new HashMap<>();
         private final Map<String, OpelAsyncFunction<?>> functions = new HashMap<>();
-        private Optional<EvalContext> parentEvalContext = Optional.empty();
+        private EvalContext parentEvalContext = fromMaps(variables, functions);
```

Comment:
variables and functions maps hear are always empty. You can write 
private EvalContext parentEvalContext =  EvalContext.empty()